### PR TITLE
Fix two bugs with concat for 0-size tensors

### PR DESCRIPTION
**Pull request by aprocter**
**Bug 1**

The reference for `concat`, and the CPU kernel emitter derived from it, were failing when the _last_ argument of `concat` had a zero-length dimension:

```
$ test/unit-test  --gtest_filter='*.concat_zero_length_1d*'
Note: Google Test filter = *.concat_zero_length_1d*
[==========] Running 4 tests from 2 test cases.
[----------] Global test environment set-up.
[----------] 2 tests from INTERPRETER
[ RUN      ] INTERPRETER.concat_zero_length_1d_last
unknown file: Failure
C++ exception with description "The start corner is out of bounds at axis 0" thrown in the test body.
[  FAILED  ] INTERPRETER.concat_zero_length_1d_last (1 ms)
[ RUN      ] INTERPRETER.concat_zero_length_1d_middle
[       OK ] INTERPRETER.concat_zero_length_1d_middle (0 ms)
[----------] 2 tests from INTERPRETER (1 ms total)

[----------] 2 tests from CPU
[ RUN      ] CPU.concat_zero_length_1d_last
unknown file: Failure
C++ exception with description "The start corner is out of bounds at axis 0" thrown in the test body.
[  FAILED  ] CPU.concat_zero_length_1d_last (1 ms)
[ RUN      ] CPU.concat_zero_length_1d_middle
[       OK ] CPU.concat_zero_length_1d_middle (1421 ms)
[----------] 2 tests from CPU (1422 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 2 test cases ran. (1423 ms total)
[  PASSED  ] 2 tests.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] INTERPRETER.concat_zero_length_1d_last
[  FAILED  ] CPU.concat_zero_length_1d_last

 2 FAILED TESTS
```

I think this is technically a bug with `CoordinateTransform` on which I should probably dig a bit deeper, but there is a pretty-obviously-correct fix for it at the concat-reference level, so that's what I've done here.

**Bug 2**

Separately, `concat` on MKLDNN crashes when given a zero-byte tensor even if it's _not_ the last argument to `concat`:

```
$ test/unit-test  --gtest_filter='*.concat_zero_length_*'
Note: Google Test filter = *.concat_zero_length_*
[==========] Running 6 tests from 2 test cases.
[----------] Global test environment set-up.
[----------] 3 tests from INTERPRETER
[ RUN      ] INTERPRETER.concat_zero_length_1d_last
[       OK ] INTERPRETER.concat_zero_length_1d_last (1 ms)
[ RUN      ] INTERPRETER.concat_zero_length_1d_middle
[       OK ] INTERPRETER.concat_zero_length_1d_middle (0 ms)
[ RUN      ] INTERPRETER.concat_zero_length_4d_middle
[       OK ] INTERPRETER.concat_zero_length_4d_middle (0 ms)
[----------] 3 tests from INTERPRETER (1 ms total)

[----------] 3 tests from CPU
[ RUN      ] CPU.concat_zero_length_1d_last
[       OK ] CPU.concat_zero_length_1d_last (1388 ms)
[ RUN      ] CPU.concat_zero_length_1d_middle
[       OK ] CPU.concat_zero_length_1d_middle (230 ms)
[ RUN      ] CPU.concat_zero_length_4d_middle
unknown file: Failure
C++ exception with description "std::exception" thrown in the test body.
[  FAILED  ] CPU.concat_zero_length_4d_middle (2 ms)
[----------] 3 tests from CPU (1620 ms total)

[----------] Global test environment tear-down
[==========] 6 tests from 2 test cases ran. (1621 ms total)
[  PASSED  ] 5 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] CPU.concat_zero_length_4d_middle

 1 FAILED TEST
```

Worked around this by not assigning zero-byte tensors to MKLDNN.

### CODE REVIEW

File: src/ngraph/runtime/reference/concat.hpp [hunk 1]
```diff
@@ -39,6 +39,24 @@ namespace ngraph
 
                 for (size_t i = 0; i < args.size(); i++)
                 {
+                    // CoordinateTransform gets confused when the last input has a zero-size dim, so we will
+                    // just skip for zero-element tensors.
+                    bool any_zero = false;
+
+                    for (size_t len : in_shapes[i])
```
Comment (rkimballn1) [inline src/ngraph/runtime/reference/concat.hpp]: I've seen this code somewhere before. You could make this a function `any_zeros(Shape)` if you feel like it. The code would read easier.

File: test/backend_test.in.cpp [hunk 2]
```diff
@@ -668,6 +668,100 @@ NGRAPH_TEST(${BACKEND_NAME}, concat_5d)
...
+
+    auto a = backend->create_tensor(element::f32, shape_a);
+    copy_data(a, a_data);
+    auto b = backend->create_tensor(element::f32, shape_b);
+    copy_data(b, b_data);
+
+    auto result = backend->create_tensor(element::f32, shape_r);
+
+    backend->call(f, {result}, {a, b});
+    EXPECT_EQ((vector<float>{1, 2, 3, 4}), read_vector<float>(result));
```
Comment (rkimballn1) [inline test/backend_test.in.cpp]: Equality comparisons with float is not good in general. Sort of ok here...

File: src/ngraph/runtime/reference/concat.hpp (hunk 1 above)
Comment (aprocter) [reply src/ngraph/runtime/reference/concat.hpp]: I'll go ahead and do that, yeah. Thinking I'll test `shape_size(in_shapes[i]) == 0`.

File: test/backend_test.in.cpp (hunk 2 above)
Comment (aprocter) [reply test/backend_test.in.cpp]: Strictly speaking all we need to test is that it doesn't crash, I suppose.

Comment (diyessi) [reply test/backend_test.in.cpp]: The equality test should be against the non-zero-shaped input, not against a duplication of its initialization.

Comment (aprocter) [reply test/backend_test.in.cpp]: That'd work for this test but would be a bit awkward for the other ones introduced in this PR.

**Comment by rkimballn1** [comment]
TESTNOW

**Pull request merged**